lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,218 @@
+require 'deep_merge/core'
+
+module Puppet::Pops
+  # Merges to objects into one based on an implemented strategy.
+  #
+  class MergeStrategy
+    TypeAsserter = Puppet::Pops::Types::TypeAsserter
+    TypeParser = Puppet::Pops::Types::TypeParser
+
+    # The type used for validation of the _merge_ argument
+    def self.merge_t
+      @@merge_t ||=  TypeParser.new.parse("Variant[String[1],Runtime[ruby,'Symbol'],Hash[Variant[String[1],Runtime[ruby,'Symbol']],Scalar,1]]")
+    end
+    private_class_method :merge_t
+
+    def self.strategies
+      @@strategies ||= {}
+    end
+    private_class_method :strategies
+
+    # Finds the merge strategy for the given key and returns it.
```

Comment:
replace "key" with "merge_strategy_key" (can be confused with merge strategy being available per data key).